Crash report 3.9-rc-1: RuntimeException in FetchBlogListAbstract

Probably a side effect of the 2 step auth merge.

```
java.lang.ExceptionInInitializerError
       at org.wordpress.android.ui.accounts.SignInFragment.refreshBlogContent(SignInFragment.java:844)
       at org.wordpress.android.ui.accounts.SignInFragment.refreshFirstBlogContent(SignInFragment.java:857)
       at org.wordpress.android.ui.accounts.SignInFragment.access$1900(SignInFragment.java:70)
       at org.wordpress.android.ui.accounts.SignInFragment$11.onSuccess(SignInFragment.java:434)
       at org.wordpress.android.ui.accounts.helpers.FetchBlogListWPOrg.getBlogList(FetchBlogListWPOrg.java:87)
       at org.wordpress.android.ui.accounts.helpers.FetchBlogListWPOrg.fetchBlogList(FetchBlogListWPOrg.java:127)
       at org.wordpress.android.ui.accounts.helpers.FetchBlogListAbstract$1.run(FetchBlogListAbstract.java:27)
Caused by: java.lang.RuntimeException: Can't create handler inside thread that has not called Looper.prepare()
       at android.os.Handler.(Handler.java:121)
       at android.os.AsyncTask$InternalHandler.(AsyncTask.java:607)
       at android.os.AsyncTask$InternalHandler.(AsyncTask.java:607)
       at android.os.AsyncTask.(AsyncTask.java:190)
       at org.wordpress.android.ui.accounts.SignInFragment.refreshBlogContent(SignInFragment.java:844)
       at org.wordpress.android.ui.accounts.SignInFragment.refreshFirstBlogContent(SignInFragment.java:857)
       at org.wordpress.android.ui.accounts.SignInFragment.access$1900(SignInFragment.java:70)
       at org.wordpress.android.ui.accounts.SignInFragment$11.onSuccess(SignInFragment.java:434)
       at org.wordpress.android.ui.accounts.helpers.FetchBlogListWPOrg.getBlogList(FetchBlogListWPOrg.java:87)
       at org.wordpress.android.ui.accounts.helpers.FetchBlogListWPOrg.fetchBlogList(FetchBlogListWPOrg.java:127)
       at org.wordpress.android.ui.accounts.helpers.FetchBlogListAbstract$1.run(FetchBlogListAbstract.java:27)
```

--- a/WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListAbstract.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListAbstract.java
@@ -21,12 +21,7 @@ public abstract class FetchBlogListAbstract {
 
     public void execute(final Callback callback) {
         mCallback = callback;
-        new Thread() {
-            @Override
-            public void run() {
-                fetchBlogList(callback);
-            }
-        }.start();
+        fetchBlogList(callback);
     }
 
     protected abstract void fetchBlogList(final Callback callback);

--- a/WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/accounts/helpers/FetchBlogListWPOrg.java
@@ -1,5 +1,6 @@
 package org.wordpress.android.ui.accounts.helpers;
 
+import android.os.AsyncTask;
 import android.webkit.URLUtil;
 
 import org.wordpress.android.analytics.AnalyticsTracker;
@@ -64,67 +65,8 @@ public class FetchBlogListWPOrg extends FetchBlogListAbstract {
         }
     }
 
-    private void getBlogList(URI xmlrpcUri, final Callback callback) {
-        XMLRPCClientInterface client = XMLRPCFactory.instantiate(xmlrpcUri, mHttpUsername, mHttpPassword);
-        Object[] params = {mUsername, mPassword};
-        try {
-            Object[] userBlogs = (Object[]) client.call("wp.getUsersBlogs", params);
-            if (userBlogs == null) {
-                // Could happen if the returned server response is truncated
-                mErrorMsgId = org.wordpress.android.R.string.xmlrpc_error;
-                callback.onError(mErrorMsgId, false, false, false, client.getResponse());
-                return;
-            }
-            Arrays.sort(userBlogs, BlogUtils.BlogNameComparator);
-            List<Map<String, Object>> userBlogList = new ArrayList<Map<String, Object>>();
-            for (Object blog : userBlogs) {
-                try {
-                    userBlogList.add((Map<String, Object>) blog);
-                } catch (ClassCastException e) {
-                    AppLog.e(T.NUX, "invalid data received from XMLRPC call wp.getUsersBlogs");
-                }
-            }
-            callback.onSuccess(userBlogList);
-            return;
-        } catch (XmlPullParserException parserException) {
-            mErrorMsgId = org.wordpress.android.R.string.xmlrpc_error;
-            AppLog.e(T.NUX, "invalid data received from XMLRPC call wp.getUsersBlogs", parserException);
-        } catch (XMLRPCFault xmlRpcFault) {
-            handleXmlRpcFault(xmlRpcFault);
-        } catch (XMLRPCException xmlRpcException) {
-            AppLog.e(T.NUX, "XMLRPCException received from XMLRPC call wp.getUsersBlogs", xmlRpcException);
-            mErrorMsgId = org.wordpress.android.R.string.no_site_error;
-        } catch (SSLHandshakeException e) {
-            if (xmlrpcUri.getHost().endsWith("wordpress.com")) {
-                mErroneousSslCertificate = true;
-            }
-            AppLog.w(T.NUX, "SSLHandshakeException failed. Erroneous SSL certificate detected.");
-        } catch (IOException e) {
-            AppLog.e(T.NUX, "Exception received from XMLRPC call wp.getUsersBlogs", e);
-            mErrorMsgId = org.wordpress.android.R.string.no_site_error;
-        }
-        callback.onError(mErrorMsgId, false, mHttpAuthRequired, mErroneousSslCertificate, client.getResponse());
-    }
-
     public void fetchBlogList(Callback callback) {
-        String xmlrpcUrl = null;
-        if (mSelfHostedUrl != null && mSelfHostedUrl.length() != 0) {
-            xmlrpcUrl = getSelfHostedXmlrpcUrl(mSelfHostedUrl);
-        }
-
-        if (xmlrpcUrl == null) {
-            if (!mHttpAuthRequired && mErrorMsgId == 0) {
-                mErrorMsgId = org.wordpress.android.R.string.no_site_error;
-            }
-            callback.onError(mErrorMsgId, false, mHttpAuthRequired, mErroneousSslCertificate, "");
-            return;
-        }
-
-        // Validate the URL found before calling the client. Prevent a crash that can occur
-        // during the setup of self-hosted sites.
-        URI xmlrpcUri;
-        xmlrpcUri = URI.create(xmlrpcUrl);
-        getBlogList(xmlrpcUri, callback);
+        (new FetchBlogListTask(callback)).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
     }
 
     private String getRsdUrl(String baseUrl) throws SSLHandshakeException {
@@ -278,5 +220,85 @@ public class FetchBlogListWPOrg extends FetchBlogListAbstract {
         }
 
         return xmlrpcUrl;
+    }
+
+    public class FetchBlogListTask extends AsyncTask<Void, Void, List<Map<String, Object>>> {
+        private final Callback mCallback;
+        private String mClientResponse = "";
+
+        public FetchBlogListTask(Callback callback) {
+            mCallback = callback;
+        }
+
+        @Override
+        protected List<Map<String, Object>> doInBackground(Void... notUsed) {
+            String xmlrpcUrl = null;
+            if (mSelfHostedUrl != null && mSelfHostedUrl.length() != 0) {
+                xmlrpcUrl = getSelfHostedXmlrpcUrl(mSelfHostedUrl);
+            }
+
+            if (xmlrpcUrl == null) {
+                if (!mHttpAuthRequired && mErrorMsgId == 0) {
+                    mErrorMsgId = org.wordpress.android.R.string.no_site_error;
+                }
+                return null;
+            }
+
+            // Validate the URL found before calling the client. Prevent a crash that can occur
+            // during the setup of self-hosted sites.
+            URI xmlrpcUri;
+            xmlrpcUri = URI.create(xmlrpcUrl);
+            XMLRPCClientInterface client = XMLRPCFactory.instantiate(xmlrpcUri, mHttpUsername, mHttpPassword);
+            Object[] params = {mUsername, mPassword};
+            try {
+                Object[] userBlogs = (Object[]) client.call("wp.getUsersBlogs", params);
+                if (userBlogs == null) {
+                    // Could happen if the returned server response is truncated
+                    mErrorMsgId = org.wordpress.android.R.string.xmlrpc_error;
+                    mClientResponse = client.getResponse();
+                    return null;
+                }
+                Arrays.sort(userBlogs, BlogUtils.BlogNameComparator);
+                List<Map<String, Object>> userBlogList = new ArrayList<Map<String, Object>>();
+                for (Object blog : userBlogs) {
+                    try {
+                        userBlogList.add((Map<String, Object>) blog);
+                    } catch (ClassCastException e) {
+                        AppLog.e(T.NUX, "invalid data received from XMLRPC call wp.getUsersBlogs");
+                    }
+                }
+                return userBlogList;
+            } catch (XmlPullParserException parserException) {
+                mErrorMsgId = org.wordpress.android.R.string.xmlrpc_error;
+                AppLog.e(T.NUX, "invalid data received from XMLRPC call wp.getUsersBlogs", parserException);
+            } catch (XMLRPCFault xmlRpcFault) {
+                handleXmlRpcFault(xmlRpcFault);
+            } catch (XMLRPCException xmlRpcException) {
+                AppLog.e(T.NUX, "XMLRPCException received from XMLRPC call wp.getUsersBlogs", xmlRpcException);
+                mErrorMsgId = org.wordpress.android.R.string.no_site_error;
+            } catch (SSLHandshakeException e) {
+                if (xmlrpcUri.getHost() != null && xmlrpcUri.getHost().endsWith("wordpress.com")) {
+                    mErroneousSslCertificate = true;
+                }
+                AppLog.w(T.NUX, "SSLHandshakeException failed. Erroneous SSL certificate detected.");
+            } catch (IOException e) {
+                AppLog.e(T.NUX, "Exception received from XMLRPC call wp.getUsersBlogs", e);
+                mErrorMsgId = org.wordpress.android.R.string.no_site_error;
+            }
+            mClientResponse = client.getResponse();
+            return null;
+        }
+
+        protected void onPostExecute(List<Map<String, Object>> userBlogList) {
+            if (userBlogList == null) {
+                mCallback.onError(mErrorMsgId, false, mHttpAuthRequired, mErroneousSslCertificate, mClientResponse);
+            } else {
+                mCallback.onSuccess(userBlogList);
+            }
+        }
+
+        protected void onCancelled() {
+            mCallback.onError(mErrorMsgId, false, mHttpAuthRequired, mErroneousSslCertificate, mClientResponse);
+        }
     }
 }

--- a/WordPress/src/main/java/org/xmlrpc/android/ApiHelper.java
+++ b/WordPress/src/main/java/org/xmlrpc/android/ApiHelper.java
@@ -4,11 +4,6 @@ import android.content.Context;
 import android.os.AsyncTask;
 import android.util.Xml;
 
-import com.android.volley.NetworkResponse;
-import com.android.volley.Request;
-import com.android.volley.ServerError;
-import com.android.volley.toolbox.RequestFuture;
-import com.android.volley.toolbox.StringRequest;
 import com.google.gson.Gson;
 
 import org.wordpress.android.WordPress;
@@ -28,20 +23,21 @@ import org.wordpress.android.util.MapUtils;
 import org.xmlpull.v1.XmlPullParser;
 import org.xmlpull.v1.XmlPullParserException;
 
+import java.io.BufferedReader;
 import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
 import java.io.StringReader;
 import java.net.HttpURLConnection;
+import java.net.URL;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Vector;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.TimeoutException;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
@@ -1051,46 +1047,30 @@ public class ApiHelper {
     }
 
     /**
-     * Synchronous method to fetch the String content at the specified URL.
+     * Synchronous method to fetch the String content at the specified HTTP URL.
      *
-     * @param url                     URL to fetch contents for.
+     * @param stringUrl URL to fetch contents for.
      * @return content of the resource, or null if URL was invalid or resource could not be retrieved.
      */
-    public static String getResponse(final String url) throws SSLHandshakeException {
-        return getResponse(url, 3);
-    }
-
-    private static String getResponse(final String url, int maxRedirection)
-            throws SSLHandshakeException {
-        RequestFuture<String> requestFuture = RequestFuture.newFuture();
-        StringRequest stringRequest = new StringRequest(Request.Method.GET, url, requestFuture, requestFuture);
-        WordPress.requestQueue.add(stringRequest);
+    public static String getResponse(final String stringUrl) throws SSLHandshakeException {
         try {
-            // Wait for the response
-            return requestFuture.get(30, TimeUnit.SECONDS);
-        } catch (InterruptedException e) {
-            AppLog.e(T.API, e);
-            return null;
-        } catch (ExecutionException e) {
-            if (e.getCause() instanceof ServerError) {
-                NetworkResponse networkResponse = ((ServerError) e.getCause()).networkResponse;
-                if ((networkResponse != null) && (networkResponse.statusCode == HttpURLConnection.HTTP_MOVED_PERM ||
-                                                  networkResponse.statusCode == HttpURLConnection.HTTP_MOVED_TEMP)) {
-                    String newUrl = networkResponse.headers.get("Location");
-                    if (maxRedirection > 0) {
-                        return getResponse(newUrl, maxRedirection - 1);
-                    }
-                }
+            URL url = new URL(stringUrl);
+            HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
+            urlConnection.setInstanceFollowRedirects(true);
+            InputStream inputStream = urlConnection.getInputStream();
+            BufferedReader r = new BufferedReader(new InputStreamReader(inputStream));
+            StringBuilder sb = new StringBuilder();
+            String line;
+            while ((line = r.readLine()) != null) {
+                sb.append(line);
             }
-            if (e.getCause() != null && e.getCause().getCause() instanceof SSLHandshakeException) {
-                throw (SSLHandshakeException) e.getCause().getCause();
-            }
+            return sb.toString();
+        } catch (SSLHandshakeException e) {
+            throw e;
+        } catch (IOException e) {
             AppLog.e(T.API, e);
-            return null;
-        } catch (TimeoutException e) {
-            AppLog.e(T.API, e);
-            return null;
         }
+        return null;
     }
 
     /**

